### DISCUSSION TITLE
PHP and NodeJS Benchmark with 512 concurrency.  How PHP wins?

### DISCUSSION THREAD

**Discussion by indramal** [TechEmpower/FrameworkBenchmarks#8170] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/28696256/233773022-4b02d4dc-0354-4e25-a6d9-8aaee304fd4f.png)

PHP + NGINX win for high concurrent connection performance than NodeJS. Is it possible? How it happened?

**Comment by joanhey** [comment, marked as answer]
In reality the difference is bigger. Because the spectre/meldown/retbleed mitigations affected the Xeon Gold in the citrine servers.
And Nginx is very affected for that. We are waiting the new servers, with new CPUs.

Php-ngx is not the normal PHP + Nginx, it's using ngx-php that is a PHP lib embedded in Nginx.
The normal php with Nginx is at number 6, and now have only 9K req/s more than Nodejs. In round 18, before the mitigations, normal PHP was ~40K req/s faster than Nodejs.

When Nodejs use the database, drop drastically.
And Workerman (php from cli without server) or Ngx-php are a lot faster than Nodejs in all tests.

Normal PHP with Nginx, using Fastcgi never will be faster than Nodejs in json and plaintext, but it's faster when using the database. And still I'll try to make it faster when normal php use the database.

And in a real app, normally Nginx will be in front of Nodejs, as a Proxy server and drop a little more.

**Reply by indramal** [reply]
@joanhey 

Thank you very much for your hard works and information. It is very helpful for large scale websites developers. 

> And in a real app, normally Nginx will be in front of Nodejs, as a Proxy server and drop a little more.

This is very true. NGINX need in front to use load balance, proxy etc. NodeJS or ExpressJS not suitable for that.

When considering coding time, NodeJS, spring boot those are much faster coding time because those have more libraries. When use PHP + Laveral, it is worst than NodeJS or Spring Boots. I believe that is why most productions use spring boot or NodeJS to reduce project time. But it will cover in cost of resources because those needs high resources to get high performance.  

I think takes times to write codes is better than use of NodeJS or spring boot because we need to pay for server resources every month.

**Reply by ishtms** [reply]
> It is very helpful for large scale websites developers.

Large scale website developer don't really care about synthetic benchmarks.
>When considering coding time, NodeJS, spring boot those are much faster coding time because those have more libraries. When use PHP + Laveral, it is worst than NodeJS or Spring Boots.

Not sure what you mean by faster coding time. I can write code faster in rust, as compared to javascript, go, or C++. It doesn't matter if something has a lot of libraries.
> I believe that is why most productions use spring boot or NodeJS to reduce project time. But it will cover in cost of resources because those needs high resources to get high performance.

I am not sure what you mean by reduced project time. You can write a nodejs program in 10 minutes, and spend next 3 hours figuring out an undefined behaviour. Comparing languages by the time it takes to write code shouldn't be in the top 10 factors for considering what to choose for a project.

> I think takes times to write codes is better than use of NodeJS or spring boot because we need to pay for server resources every month.

?

**Reply by indramal** [reply]
> Large scale website developer don't really care about synthetic benchmarks.

If large scale web site not care about synthetic benchmarks, what is the method to choose best language, server etc.?

> Not sure what you mean by faster coding time.

When have lot of libraries or packages, it will care about most of coding part so we only need write less code then coding time will reduce. Someone have a good knowledge and practice about some language. Example: Rust but when that person have a library in Rust, code is less then much faster coding time. Library use or not depend on person. I mean even if someone expert to some language, libraries can reduce coding time. So NodeJS, spring boot have lot of libraries i believe that is why most of deveopers use those two. 

> top 10 factors

Actually, I mention it because I need to tell "needs high resources to get high performance." for NodeJs and spring boot. What factors do you think? 

> ?

Considering costing side, PHP use is better than NodeJS or spring boot because we need to pay much cost for server resources every month. (needs high resources to get high performance.) 

**NOTE:** All above details write considering details in above photo.

**Reply by ishtms** [reply]
> If large scale web site not care about synthetic benchmarks, what is the method to choose best language, server etc.?

Facebook runs on PHP. Considering that, everyone should start using PHP ins't it? You don't choose the best language, there is no best language or best framework. Choose something that can get your job done quickly. You can profile your code and fill in the gaps.
> So NodeJS, spring boot have lot of libraries i believe that is why most of deveopers use those two.

This is untrue. People do not use nodejs because it has a lot of libraries. People use it for the convinience of using one langauge for both the frontend and the backend, as well as mobile/gui etc. On the other hand, same for spring boot - I don't think there's a lot of "I/O intensive" apps being shipped with spring boot.

> Actually, I mention it because I need to tell "needs high resources to get high performance." for NodeJs and spring boot. What factors do you think?

What kind of resources are you talking about? Nodejs is a production grade software, still used by unicorns. Nodejs is pretty performant out of the box. It's much faster than you actually think. It uses libuv, a c++ implementation behind the scene. You're making general assumptions which are totally incorrect and flawed.

> Considering costing side, PHP use is better than NodeJS or spring boot because we need to pay much cost for server resources every month. (needs high resources to get high performance.)

I have several nodejs apps shipped on AWS serving a million requests every week, and I do not pay anything as far except EC2's monthly t2-micro costs. They run perfectly find on the smallest ec2 instance AWS offers. Maybe consider writing better code instead of finding "the best" framework/language? Someone can write a server in golang and it would still perform bad compared to a similar of javascript if they do not know how should things be done in golang.

**Reply by joanhey** [reply]
Anybody can use any language, it they don't know and write bad code is the same the language.
But still isn't the same use one language or another.

Anybody can make a fast language or framework slow, but nobody can make an slow language or framework faster, without touch the base core code.

.Net have a good example, after optimize for this benchmark.
And yes the cost is important !!!

Read this, from 40K cores to 20K.
https://devblogs.microsoft.com/dotnet/azure-active-directorys-gateway-service-is-on-net-core-3-1/

Also, https://devblogs.microsoft.com/dotnet/announcing-asp-net-core-in-net-5/

A benchmark is NOT a competition, is a very good tool to optimize  the languages and frameworks.

**Reply by joanhey** [reply]
Still .Net core is fast as platform, writing code by good experts.
But when you use .Net MVC is very slow.

**Reply by indramal** [reply]
> What kind of resources are you talking about? Nodejs is a production grade software, still used by unicorns. Nodejs is pretty performant out of the box. It's much faster than you actually think. It uses libuv, a c++ implementation behind the scene. You're making general assumptions which are totally incorrect and flawed.

I mentioned hardware resources. PHP, GoLang need low resources but NodeJS and spring boot need high resources server computers. Please check hardware resources usages for PHP, GoLang, NodeJS and spring boot language for same job with optimize code. I am not telling you are wrong NodeJS very good, but I am doing comparison. 

> both the frontend and the backend, as well as mobile/gui etc. 

That is also another way telling about libraries. Frontend - React,  Angular. Backend - ExpressJS. Mobile - React Native. All these are libraries. If not available those, will reduce developers. Considering spring boot, backend can develop in less code.

> I have several nodejs apps shipped on AWS serving a million requests every week, and I do not pay anything as far except EC2's monthly t2-micro costs. They run perfectly find on the smallest ec2 instance AWS offers. Maybe consider writing better code instead of finding "the best" framework/language? Someone can write a server in golang and it would still perform bad compared to a similar of javascript if they do not know how should things be done in golang.

Which OS do you use for server? AWS Linux or Ubuntu. 

Your information correct. But I talking about this benchmark website information. Considering website details, need to assume developer has best knowledge and good written optimize code. 

> You don't choose the best language, there is no best language or best framework. Choose something that can get your job done quickly.

You are correct when consider single person or develop a backend to handing over to client and then client continued to pay hardware resources. When I hire someone to develop website backend for the own company project, that person has the best knowledge and ability to optimization the code. With that of assumption, I think need to hire PHP or GoLang experts because those can run low hardware resources servers than NodeJS and spring boot. Choose something job done quickly can valid for small projects not for large projects. If large projects run long time with high hardware resources, company get high bill. 

That is why benchmarks need to get idea about language with combination of servers and libraries. 

Also one more thing to mention, when use NodeJS or GoLang, need NGINX server to load balance for large websites. NodeJS + NGINX combination will reduce performance.

**Comment by AadaEa** [comment]
Not what I'm observing in real life though, PHP really feels faster for development, is more lightweight and simpler, which is probably why it's so much more popular than nodejs. Both have all the libraries they need to do more or less anything.

**Comment by masfahru** [comment]
Currently, there is uWebsockets.js in NodeJS which can stand up against other popular framework.
I have tested uWebsockets.js, Go Fiber, and PHP-Ngx using 3 VMs (client, database, and server, each has 8 vcpus / 8 GB memory) in Google Cloud.
This is the result:
![image](https://github.com/TechEmpower/FrameworkBenchmarks/assets/92277627/727a5a45-fbe4-41bf-9eb9-1bd227762ea0)
https://www.techempower.com/benchmarks/#section=test&shareid=3e23be64-936d-43bb-a879-2d8de884592b&test=composite

In that test, i use different database driver than the one used in this repository.

**Reply by ishtms** [reply]
uWebsockets isn't written in JS. It's using FFI to call C++ functions in the background. This is the original repo - https://github.com/uNetworking/uWebSockets
uWebsockets.js is only using C++ bindings, nothing else. Just like just-js.

**Reply by ishtms** [reply]
Again, you need to not worry too much about things, most of the frameworks here include unrealistic optimisation that really no one does in a real world scenario.

**Reply by joanhey** [reply]
I can create any local bench that is favorable to any framework.
Please add the changes to this benchmark.
Most include unrealistic optimizations, but not all !!

**Reply by joanhey** [reply]
uWebsockets.js don't have very good numbers in this benchmark.
Please add the changes.
Thank you.

**Reply by masfahru** [reply]
> Again, you need to not worry too much about things, most of the frameworks here include unrealistic optimisation that really no one does in a real world scenario.

Sorry, i don't know how many company use uWebsockets.js in their production setup, but i do know there are people who use [Hyper-Express](https://github.com/kartikk221/hyper-express) (which is powered by [uWebSockets.js](https://github.com/uNetworking/uWebSockets.js/)) by checking the weekly downloads metric in [it's npm page](https://www.npmjs.com/package/hyper-express).
For my case, i use Hyper-Express as the platform of [NestJS](https://docs.nestjs.com/fundamentals/platform-agnosticism), since Hyper-Express has Express.js API Compatibility, it doesn't need a lot of works to swap Express.js with Hyper-Express. Also, I have pushed [a test for Hyper-Express](https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/JavaScript/hyperexpress), hopefully it will get a good benchmark in the next runs / round.

**Reply by masfahru** [reply]
> uWebsockets.js don't have very good numbers in this benchmark. Please add the changes. Thank you.

Can i push an update for [uwebsockets.js](https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/JavaScript/uwebsockets.js) ?

**Reply by joanhey** [reply]
Of course, this benchmark is completely open.
Send your PR.

Update: you can also ask the maintainer of uwebsocket.js to approve the changes in the same PR.

**Reply by masfahru** [reply]
Thanks @joanhey , please review this [PR](https://github.com/TechEmpower/FrameworkBenchmarks/pull/8332)
